slh/slh_dsa.c

```diff
@@ -563,39 +577,94 @@ size_t slh_do_sign( slh_ctx_t *ctx, uint8_t *sig, const uint8_t *digest)
     return sig_sz;
 }
 
-size_t slh_sign(uint8_t *sig, const uint8_t *m, size_t m_sz,
-                const uint8_t *sk, int (*rbg)(uint8_t *x, size_t xlen),
-                const slh_param_t *prm)
+// Deterministic portion of signing
+size_t slh_sign_internal(uint8_t *sig, const uint8_t *pre, size_t pre_sz, 
+    const uint8_t *m, size_t m_sz, const uint8_t *sk,
+    const slh_param_t *prm, uint8_t *add_rnd)
 {
-    slh_ctx_t   ctx;
     uint8_t opt_rand[SLH_MAX_N];
     uint8_t digest[SLH_MAX_M];
+    slh_ctx_t   ctx;
 
     //  set up secret key etc
     prm->mk_ctx(&ctx, NULL, sk, prm);
 
-#ifdef SLH_DETERMINISTIC
-    memcpy(opt_rand, ctx.pk_seed, prm->n);
-#else
-    rbg(opt_rand, prm->n);
-#endif
+    if(deterministic_g == SLH_DETERMINISTIC)
+    {
+        memcpy(opt_rand, ctx.pk_seed, prm->n);
+    }
+    else if(deterministic_g == SLH_NON_DETERMINISTIC)
+    {
+        memcpy(opt_rand, add_rnd, prm->n);
+    }
+    else
+    {
+        return 0;    
+    }
 
     //  randomized hashing; R
     uint8_t *r  = sig;
     size_t  sig_sz = prm->n;
-    prm->prf_msg(&ctx, r, opt_rand, m, m_sz);
-    prm->h_msg(&ctx, digest, r, m, m_sz);
+    prm->prf_msg(&ctx, r, opt_rand, pre, pre_sz, m, m_sz);
+    prm->h_msg(&ctx, digest, r, pre, pre_sz, m, m_sz);
 
     //  create FORS and HT signature parts
     sig_sz += slh_do_sign(&ctx, sig + sig_sz, digest);
 
     return sig_sz;
 }
 
+// Pure signing wrapper function
+size_t slh_sign(uint8_t *sig, const uint8_t *m, size_t m_sz,
+    const uint8_t *sk, int (*rbg)(uint8_t *x, size_t xlen),
+    const slh_param_t *prm, uint8_t *ctx_str, size_t ctx_str_len)
+{
+    if (ctx_str_len > SLH_MAX_CTX_STR_LEN)
+    {
+        return 0;
+    }
+
+    uint8_t add_rnd[SLH_MAX_N];
+
+    if(deterministic_g == SLH_DETERMINISTIC)
+    {
+        // add_rnd not needed here so is uninitialized
+    }
+    else if(deterministic_g == SLH_NON_DETERMINISTIC)
+    {
+        if (rbg(add_rnd, prm->n) != 0)
+        {
+            return 0;
+        }
+        // printf("add_rnd = \r\n");
+        // for(int i = 0; i<16; i++){
+        //     printf("%d ",add_rnd[i]);
+        // }
+        // printf("\r\n");
+    }
+    else
+    {
+        return 0;
+    }
+
+    uint8_t pre[MAX_PRE_SIZE];
+    size_t pre_sz = 2 + ctx_str_len;
+
+    pre[0] = 0;
+    pre[1] = ctx_str_len;
+    for (uint8_t i = 0; i < ctx_str_len; i++)
```

Comment:
```suggestion
    for (size_t i = 0; i < ctx_str_len; i++)
```
Let's use `size_t` here to make sure that the computation doesn't wrap around and cause an infinite loop. We can get away with `uint8_t` here since `ctx_str_len` is checked to be at most 255, but it's more clear just to use `size_t`.